### ISSUE TITLE
Consider giving molds a `:type-vector` property to support auto composability

### ISSUE THREAD

**Issue by kat-co** [ag91/moldable-emacs#11]
Sorry, this got a little long. Writing this up helped me think through things.

**TLDR**: If we give mold definitions an ordered list of type information, in increasing specificity, and write a few functions, moldable-emacs can synthesize composable molds without having to write them.

---
From what I can tell, the "moldable" concept is about defining a graph:

- The vertices in the graph are states data, (i.e. a buffer) can be in
- The edges in the graph are the molds, i.e. functions which transform data from state, i.e. Mold = f(S<sub>n</sub>) = S<sub>m</sub>
- The graph is directed and permits loops, i.e. you can go from `list` -> `csv` -> `list`, or S<sub>n</sub> -> S<sub>m</sub> -> S<sub>n</sub>.

My impression is that the power of the moldable concept comes from the fact that if you can connect novel data in a buffer with a vertex in the graph with high valency (i.e. a high number of edges), you can now leverage the full body of molds to do interesting things. That is to say, if you can make the data in your buffer look like something the moldable graph knows a lot about, you can transform that buffer into interesting things.

Currently, `moldable-emacs` has the concept of "composable molds". This is a manual definition of a path through the graph. Molds already define their requisite starting state, S<sub>n</sub> which includes assertions about runtime requirements as well as data type information. If they separated out the runtime-requirements from statements about the data's state, and then declared the state they will leave a buffer in, S<sub>m</sub>, it should be possible to write an algorithm to automatically discover what "end states" are possible for the current buffer.

This would reduce the maintenance burden of manually defining every path through the graph possible, and might surprise the user at what's possible with their current buffer.

### Representing the vertices beginning state, S<sub>n</sub>
I think there are a few pieces of information:
- What libraries must be available
- Any preconditions on activation
- The most general type of data that can be accepted (I'll expound in the next section)

### Representing the vertices end state, S<sub>m</sub>
I think it should provide the type information of the data in increasing specificity (e.g. this is an `emacs-lisp` buffer, which means it's code, which means it's a tree, but it might be a tree with a pre-defined type, like an alist or plist, etc.). This is a vector of type-information, i.e. (using the previous example) <emacs-lisp, code, tree, plist>, or **TV**=&lt;T<sub>0</sub>, T<sub>1</sub>, ..., T<sub>n</sub>>. Note that the mold is in a unique position to have the highest resolution of what type-vector the data it produces has.

### Building the graph, G
When the hypothetical algorithm tries to build the graph, it would need to walk the list of known molds, V={M<sub>0</sub>, M<sub>1</sub>, ..., M<sub>n</sub>}, and for each mold, M<sub>i</sub>:
- Determine if M<sub>i</sub> is activatable, e.g. does the running emacs process contain the necessary libraries to execute the mold. This is a filter for whether the mold enters into the graph.
- Compare against each mold, M<sub>j</sub> such that f(M<sub>i</sub>, M<sub>j</sub>) = { if T<sub>j</sub> ∈ **TV<sub>i</sub>**, then T<sub>j</sub>, otherwise nil.
- If T<sub>j</sub> is returned, then we add the edge between M<sub>i</sub> and M<sub>j</sub>.

### Using the graph, G
- When a user invokes `me-mold`, walk the graph to find the vertex whose type-vector fits the buffer/data in the most general way, i.e. f(buffer, G) = M<sub>i</sub> where:
  -  f(M<sub>i</sub>) = required type T<sub>i</sub>
  -  f(buffer) = **TV<sub>b</sub>**
  - T<sub>i</sub> ∈ **TV<sub>b</sub>**
  - f(T, **TV**) = index of T, **TV<sub>n</sub>**
  - n < any other **TV<sub>n</sub>**, i.e. the match with the lowest index wins
    - In the case where there are multiple **TV<sub>n</sub>** which are equal, prefer the **TV** with the lowest magnitude (length) **| TV |**.
      - In the case where there are multiple **TV** of equal magnitude, pick the first one sorted alphanumerically (this makes things deterministic).
- Calculate the minimum spanning tree for M<sub>i</sub> (for performance, these could be cached when the graph is constructed, and when molds are added at runtime).
- Present the terminal vertices as Mold options to run.
- Traversing the path through the graph may fail if one of the mold's prerequisites fails. When this happens, a copy of the graph should be made for the buffer, the edge that failed should be removed, and the minimum spanning tree should be recalculated.

The reason to use the mold with the most general matching type-vector as a starting point is so that we get the most paths through the graph as possible. Going back to our example, <emacs-lisp, code, tree, plist>, there might be some molds that can operate on plists, but not trees. If we pick the plist mold as a starting point, we are now disconnected from any other vertices that might know what to do with emacs-lisp, code, or trees.

The reason to use the mold with the lowest magnitude type-vector is because between two molds that have matched the buffer's type in equally general ways, the mold with the higher length type-vector is making the most specific claim about the data in the buffer, and we cannot be sure the assertion holds since a mold did not generate the data. E.g. a buffer is highly likely to be <emacs-lisp, code, tree>, but it is as of yet unknown whether it is <emacs-lisp, code, tree, plist>.

### Implications

This approach changes molds a little. Instead of defining molds as BeginStateToEndState, e.g. "CSVToOrgTable", molds should only be named to refer to their end state, e.g. "ToOrgTable" (or I suggest, "to-org-table"). This is because that when the user is presented with options, the options will be terminal vertices, but the span between the user's buffer and the terminal vertex might be long. It would not make sense to see "CSVToOrgTable" if I'm looking at an emacs-lisp buffer. It would make sense (albeit maybe surprising!) to see "ToCSV".

Molds should become as small as possible to promote reuse.

Traversing the path through the graph may fail if one of the mold's prerequisites fails. For this reason, prerequisites should be used sparingly.

Composable molds go away, because moldable-emacs handles that messy bit for us :)

**Comment by ag91** [comment]
> This would reduce the maintenance burden of manually defining every path through the graph possible, and might surprise the user at what's possible with their current buffer.

I find this pretty interesting. The thing that I like is that even bare-metal as moldable-emacs is now, you can still discover paths one edge/mold at the time. 
Have you thought about the UI of your vision? When you compute all the possible compositions, as a user how can I find out the one I want? From a generated name?

Currently I list the available molds as a list of keys. I was thinking to move towards letting the user choose a mold by computing it and show the result, but maybe you got some good ideas in this regard?

**Comment by ag91** [comment]
> it would need to walk the list of known molds, V={M0, M1, ..., Mn}, and for each mold, Mi: ...
I think we could make use of a principle here: **the user is always in the lead**. The idea is that this is a tool to help you telling your story better.

Calculating all combinations of molds could become expensive pretty quickly, so if we have such a principle we could make use of history and reduce the problem to precomputing only combinations of molds that are used often?

**Comment by ag91** [comment]
> This approach changes molds a little. Instead of defining molds as BeginStateToEndState, e.g. "CSVToOrgTable", molds should only be named to refer to their end state, e.g. "ToOrgTable" (or I suggest, "to-org-table"). This is because that when the user is presented with options, the options will be terminal vertices, but the span between the user's buffer and the terminal vertex might be long. It would not make sense to see "CSVToOrgTable" if I'm looking at an emacs-lisp buffer. It would make sense (albeit maybe surprising!) to see "ToCSV".

This sounds another refactoring that I should do anyway! "to-org-table" is more pleasing than "ToOrgTable"? I am neutral since I just moved from Scala to Clojure XD 
Very good idea, thanks for sharing it.

**Comment by ag91** [comment]
> Molds should become as small as possible to promote reuse.

Any suggestions about this is welcome! I tend to make private molds first and add to core.el and contrib.el only those that I feel are very reusable or exemplar. It would be useful to get examples of good small molds and of too big molds among those that exist from you.
